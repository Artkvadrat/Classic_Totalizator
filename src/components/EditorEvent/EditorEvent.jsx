import React, { useEffect } from 'react';
import { Skeleton } from 'antd';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory, useParams } from 'react-router-dom';
import moment from 'moment';

import {
  loadData,
  saveEvent,
  changeFieldEvent
} from '../../ducks/editorEvent/editorEvent';
import styles from './EditorEvent.module.css';

const EditorEvent = () => {
  const { event, isLoading } = useSelector((state) => state.editorEvent);
  const { startTime, margin } = event;
  const dispatch = useDispatch();
  const history = useHistory();
  const { id } = useParams();

  useEffect(() => {
    dispatch(loadData(id));
  }, [dispatch, id]);

  const handleChange = (event) => {
    const fieldName = event.target.name;
    const fieldValue =
      fieldName === 'margin' ? +event.target.value : event.target.value;

    dispatch(changeFieldEvent(fieldName, fieldValue));
  };

  const submit = (e) => {
    e.preventDefault();

    if (+margin > 0 && moment(startTime) > Date.now()) {
      dispatch(
        saveEvent({ ...event, startTime: moment(startTime).format() }, id)
      ).then(history.push('/'));
    }
  };

  return isLoading ? (
    <Skeleton active />
  ) : (
    <form className={styles['form-editor']} onSubmit={submit}>
      <span className={styles['title-value']}>Change event</span>

      <span>Enter the margin, %</span>
      <input
        className={styles['margin-value']}
        type="number"
        name="margin"
        value={margin}
        onChange={handleChange}
      />

      <span>Enter the date of event</span>
      <input
        className={styles['start-data']}
        type="datetime-local"
        name="startTime"
        value={startTime}
        required
        onChange={handleChange}
      />

      <input
        className={styles['create-action']}
        type="submit"
        value="Save event"
      />
    </form>
  );
};
export default EditorEvent;
