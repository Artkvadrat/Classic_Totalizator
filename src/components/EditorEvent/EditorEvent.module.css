.form-editor {
    max-width: 500px;
    background-color: #CFCFCF;
    margin: 50px auto;
    padding: 15px;
}
input {
    outline: none;
    padding: 5px;
    border-radius: 4px;
    border: 1px solid #000;
}
span,
.sport-name,
.margin-value,
.create-action {
    display: block;
}
.title-value {
    text-align: center;
    margin-bottom: 10px;
    font-size: 25px;
    font-weight: 700;
}

.margin-value,
.start-data,
.start-time {
    margin: 5px 0 10px;
}

.start-data {
    margin-right: 10px;
}

.create-action {
    min-width: 150px;
    margin: 35px auto 0;
    background-color: blue;
    cursor: pointer;
    color: #fff;
}
.create-action:hover {
    background-color: green;
}
