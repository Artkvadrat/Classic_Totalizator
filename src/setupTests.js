import enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-17-updated';

enzyme.configure({ adapter: new Adapter() });

const localStorageMock = {
  getItem: jest.fn(),
  setItem: jest.fn(),
  removeItem: jest.fn(),
  clear: jest.fn()
};
window.localStorage = localStorageMock;
