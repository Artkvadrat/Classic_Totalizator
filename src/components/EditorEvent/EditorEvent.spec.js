import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { shallow, mount } from 'enzyme';
import { useHistory, useParams } from 'react-router-dom';
import {
  loadData,
  saveEvent,
  changeFieldEvent,
  clearEvent
} from '../../ducks/editorEvent/editorEvent';

import '../../setupTests';

import EditorEvent from './EditorEvent';

jest.mock('react-router-dom', () => ({
  useHistory: jest.fn(),
  useParams: jest.fn()
}));

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn()
}));

jest.mock('../../ducks/editorEvent/editorEvent', () => ({
  loadData: jest.fn(),
  saveEvent: jest.fn(),
  changeFieldEvent: jest.fn(),
  clearEvent: jest.fn()
}));

describe('EditorEvent component', () => {
  const history = {
    push: jest.fn()
  };
  const dispatch = jest.fn();

  beforeEach(() => {
    useDispatch.mockReturnValue(dispatch);
    useHistory.mockReturnValue(history);
    loadData.mockReturnValue(Symbol.for('load'));
    changeFieldEvent.mockReturnValue(Symbol.for('change'));
    clearEvent.mockReturnValue(Symbol.for('clear'));
    
    useSelector.mockReturnValue({
      event: {
        id: '3fa85f64-5717-4562-b3fc-2c963f66afa6',
        margin: 5,
        startTime: '2021-05-24T04:55:03'
      }
    });
    jest.clearAllMocks();
  });

  it('should dispatch loadData action on mount', () => {
    const params = { id: '3fa85f64-5717-4562-b3fc-2c963f66afa6' };
    useParams.mockReturnValue(params);
    expect(dispatch).not.toHaveBeenCalled();
    expect(loadData).not.toHaveBeenCalled();

    const result = mount(<EditorEvent />);

    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith(Symbol.for('load'));
    expect(loadData).toHaveBeenCalledTimes(1);
    expect(loadData).toHaveBeenCalledWith(params.id);
  });
  it('should render form', () => {
    const result = shallow(<EditorEvent />);
    const form = result.find('.form-editor');
    expect(form.exists()).toEqual(true);
    expect(result).toMatchSnapshot();
  });
  it('should save event on submit form', () => {
    const result = shallow(<EditorEvent />);
    const fakeEvent = {
      preventDefault: jest.fn()
    };

    const form = result.find('.form-editor');
    form.simulate('submit', fakeEvent);

    expect(fakeEvent.preventDefault).toHaveBeenCalled();
    expect(saveEvent).toHaveBeenCalledTimes(1);
    expect(saveEvent).toHaveBeenCalledWith({
      id: '3fa85f64-5717-4562-b3fc-2c963f66afa6',
      margin: 5,
      startTime: '2021-05-24T04:55:03'
    });
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith(Symbol.for('clear'));
    expect(clearEvent).toHaveBeenCalledTimes(1);
    expect(history.push).toHaveBeenCalledTimes(1);
  });
  it("shouldn't save event on submit form", () => {
    useSelector.mockReturnValue({
      event: {
        id: '3fa85f64-5717-4562-b3fc-2c963f66afa6',
        margin: 0,
        startTime: '2021-05-24T04:55:03'
      }
    });

    const result = shallow(<EditorEvent />);
    const fakeEvent = {
      preventDefault: jest.fn()
    };
    const form = result.find('.form-editor');
    form.simulate('submit', fakeEvent);

    expect(fakeEvent.preventDefault).toHaveBeenCalled();
    expect(saveEvent).not.toHaveBeenCalled();
    expect(dispatch).not.toHaveBeenCalled();
    expect(clearEvent).not.toHaveBeenCalled();
    expect(history.push).not.toHaveBeenCalled();
  });
  it('should dispatch changeFieldEvent action on change field form', () => {
    const result = shallow(<EditorEvent />);
    const fakeEventMargin = {
      target: {
        name: 'margin',
        value: '4'
      }
    };
    const fakeEventDate = {
      target: {
        name: 'startTime',
        value: '2021-05-23T11:18:21'
      }
    };

    const margin = result.find('.margin-value');
    margin.simulate('change', fakeEventMargin);
    const date = result.find('.start-data');
    date.simulate('change', fakeEventDate);

    expect(dispatch).toHaveBeenCalledTimes(2);
    expect(dispatch).toHaveBeenCalledWith(Symbol.for('change'));
    expect(changeFieldEvent).toHaveBeenCalledTimes(2);
    expect(changeFieldEvent).toHaveBeenNthCalledWith(1, 'margin', 4);
    expect(changeFieldEvent).toHaveBeenNthCalledWith(
      2,
      'startTime',
      '2021-05-23T11:18:21'
    );
  });
});
