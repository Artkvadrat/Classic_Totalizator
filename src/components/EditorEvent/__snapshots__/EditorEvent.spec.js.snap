// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorEvent component should render form 1`] = `
<form
  className="form-editor"
  onSubmit={[Function]}
>
  <span
    className="title-value"
  >
    Edit event
  </span>
  <span>
    Вкажіть маржу, %
  </span>
  <input
    className="margin-value"
    name="margin"
    onChange={[Function]}
    type="number"
    value={5}
  />
  <span>
    Вкажіть дату та час початку події
  </span>
  <input
    className="start-data"
    name="startTime"
    onChange={[Function]}
    required={true}
    type="datetime-local"
    value="2021-05-24T04:55:03"
  />
  <input
    className="create-action"
    type="submit"
    value="Зберегти зміни"
  />
</form>
`;
